### DISCUSSION TITLE
Supporting NSCoding

### DISCUSSION THREAD

**Discussion by amomchilov** [trilemma-dev/SecureXPC#79] [Show and tell]
I thought you might find this interesting, so I'm posting it here as an FYI:

https://stackoverflow.com/a/71279637/3141234

**Comment by jakaplan** [comment]
Thanks for sharing. This sounds consistent with what we discussed in the past that `Codable` is a good fit for value types like `struct` and `enum`, but doesn't play nicely with arbitrary `class`es.

For what it's worth, I prefer `Codable`'s semantics. Having cycles in serialized formats seems like it's just asking for trouble considering the differences that exist between languages and runtimes.

**Reply by amomchilov** [reply]
ehhhhhhhh

I don't prefer classes, but there are certainly times when reference semantics are useful. Arbitrary graphs structures are inevitable, they have more expressive power to model the real world than limiting yourself to only acyclic graphs (trees).

The rise of value-semantic serialization formats (like JSON) has destroyed the support for reference semantics. But since the need hasn't gone away, it forced a lot of people to reinvent reference semantics in their own unique, non-interoperabile, typically-jank way. E.g. you'll see JSON payloads with an "all objects" dictionary, that contains a dictionary of every object in the payload, keyed by some handle/id. Then, every object reference is serialized using these handles, rather than "proper" object references. Then it becomes a client responsibility to follow all these handles and do look-ups against the "all objects" dictionary. It all becomes a mess :/

**Reply by amomchilov** [reply]
Update: Turns out archiving NSImages (particularly app icons, which was my use-case) was a really bade idea. It eagerly loads, decompresses and serializes *all* the underlying image representations, which means a simple app icon blows up to as much as 37 MB!

**Reply by jakaplan** [reply]
That it has a large in-memory representation isn't too surprising to me. But that it chooses to serialize itself so inefficient is quite unexpected. Seems rather un-Apple like; they're normally very thoughtful in terms of efficient resource usage whether that be memory, storage, compute, etc.

**Reply by amomchilov** [reply]
TBF, I don't think there's any universally good way to do it. If they don't serialize all the representations, then the scaling on the "other end" would be bad, after deserialization
